torch/csrc/api/include/torch/data/samplers/base.h

```diff
@@ -0,0 +1,29 @@
+#pragma once
+
+#include <ATen/optional.h>
+
+#include <cstddef>
+#include <vector>
+
+namespace torch {
+namespace data {
+namespace samplers {
+
+/// A `Sampler` is an object that yields indices with which to index into a
+/// dataset.
+class Sampler {
+ public:
+  virtual ~Sampler() = default;
+
+  /// Resets the `Sampler`'s internal state.
+  /// Typically called before a new epoch.
+  virtual void reset() = 0;
```

Comment:
Could we add a method to set_size for the sampler. This will let us do sampling on prefetched data with hierarchical sampling. 
`virtual void set_size(size_t size) = 0;`